subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/LinuxFileWatcherRegistryFactory.java

```diff
@@ -41,6 +46,39 @@ protected LinuxFileWatcher createFileWatcher(BlockingQueue<FileWatchEvent> fileE
 
     @Override
     protected FileWatcherUpdater createFileWatcherUpdater(LinuxFileWatcher watcher, FileWatcherProbeRegistry probeRegistry, WatchableHierarchies watchableHierarchies) {
-        return new NonHierarchicalFileWatcherUpdater(watcher, probeRegistry, watchableHierarchies);
+        return new NonHierarchicalFileWatcherUpdater(watcher, probeRegistry, watchableHierarchies, new LinuxMovedWatchedDirectoriesSupplier(watcher, watchableHierarchies));
+    }
+
+    private static class LinuxMovedWatchedDirectoriesSupplier implements AbstractFileWatcherUpdater.MovedWatchedDirectoriesSupplier {
+        private final LinuxFileWatcher watcher;
+        private final WatchableHierarchies watchableHierarchies;
+
+        public LinuxMovedWatchedDirectoriesSupplier(LinuxFileWatcher watcher, WatchableHierarchies watchableHierarchies) {
+            this.watcher = watcher;
+            this.watchableHierarchies = watchableHierarchies;
+        }
+
+        @Override
+        public Collection<File> stopWatchingMovedPaths(SnapshotHierarchy vfsRoot) {
+            List<File> pathsToCheck = new ArrayList<>();
+            vfsRoot.rootSnapshots().forEach(snapshot -> {
+                if (watchableHierarchies.shouldWatch(snapshot)) {
+                    switch (snapshot.getType()) {
+                        case RegularFile:
+                            pathsToCheck.add(new File(snapshot.getAbsolutePath()).getParentFile());
+                            break;
+                        case Directory:
+                            pathsToCheck.add(new File(snapshot.getAbsolutePath()));
+                            break;
+                        case Missing:
+                            // Ignore
+                            break;
+                        default:
+                            throw new IllegalArgumentException("Unexpected file type:" + snapshot.getType());
+                    }
+                }
+            });
```

Comment:
How about?

```suggestion
            Collection<File> pathsToCheck = vfsRoot.rootSnapshots()
                .filter(snapshot -> snapshot.getType() != FileType.Missing)
                .filter(watchableHierarchies::shouldWatch)
                .map(snapshot -> {
                    switch (snapshot.getType()) {
                        case RegularFile:
                            return new File(snapshot.getAbsolutePath()).getParentFile();
                        case Directory:
                            return new File(snapshot.getAbsolutePath());
                        default:
                            throw new IllegalArgumentException("Unexpected file type:" + snapshot.getType());
                    }
                })
                .collect(ImmutableList.toImmutableList());
```

We could also not wrap things in a `File`.